### ISSUE TITLE
regression in 1.12.0: un-suppressable error with no line number when subtyping dataclass when field has incompatible subtype

### ISSUE THREAD

**Issue by DetachHead** [python/mypy#18115]
```py
from dataclasses import dataclass


@dataclass
class Foo:
    a: object


@dataclass
class Bar(Foo):
    a: int
```
```
main.py: error: Argument 1 of "__replace__" is incompatible with supertype "Foo"; supertype defines the argument type as "object"  [override]
main.py: note: This violates the Liskov substitution principle
main.py: note: See https://mypy.readthedocs.io/en/stable/common_issues.html#incompatible-overrides
Found 1 error in 1 file (checked 1 source file)
```
[playground](https://mypy-play.net/?mypy=master&python=3.13&flags=show-error-codes%2Cstrict%2Ccheck-untyped-defs%2Cdisallow-any-decorated%2Cdisallow-any-expr%2Cdisallow-any-explicit%2Cdisallow-any-generics%2Cdisallow-any-unimported%2Cdisallow-incomplete-defs%2Cdisallow-subclassing-any%2Cdisallow-untyped-calls%2Cdisallow-untyped-decorators%2Cdisallow-untyped-defs%2Cno-implicit-reexport%2Clocal-partial-types%2Cwarn-redundant-casts%2Cwarn-return-any%2Cwarn-unreachable%2Cwarn-unused-configs%2Cwarn-unused-ignores&gist=b9355e6a2549495727a8c43c497e163c)

this error has no line number and therefore cannot be ignored. this is a big problem because not being able to `type:ignore` the error is preventing us from updating mypy

the error also occurs when the dataclass is frozen, in which case the error is a false positive here because the type of the field should be considered covariant:

```py
from dataclasses import dataclass


@dataclass(frozen=True)
class Foo:
    a: object


@dataclass(frozen=True)
class Bar(Foo):
    a: int
```

(you could argue that the error is technically still correct because the generated `__replace__` method takes the constructor arguments in a contravariant position, but `dataclasses.replace` is completely untyped anyway, so this additional safety has no benefit).

related issue on pyright: https://github.com/microsoft/pyright/issues/9351

**Comment by brianschubert** [comment]
This error seems to only occur when running mypy with Python 3.13. Checking the same code with Python 3.12 produces no error for me (both locally and on the playground).

**Comment by JukkaL** [comment]
Fixed by #18122.
